Modify rule S6021: Expand to cover unordered containers (#401)
SonarSource/rspec: rules/S6021/cfamily/rule.adoc

> Heterogeneous containers were introduced in {cpp}14 to increase performances when working with ``++std::map++``, ``++std::set++``, ``++std::multimap++``, ``++std::multiset++``, and since {cpp}20, their ``++unordered_++` counterparts.

Reviewer: ```suggestion
Heterogeneous containers were introduced in {cpp}14 to increase performance when working with ``++std::map++``, ``++std::set++``, ``++std::multimap++``, ``++std::multiset++``, and since {cpp}20, their ``++unordered_++`` counterparts.
```

>   std::string myData;

Reviewer: There's a typo two lines up but I cannot create a suggestion for it...
```
// Other functions to make the class behaves correctly
```
should be
```
// Other functions to make the class behave correctly
```
(no *s* at the end of *behave*)

In this case, ``++str++`` will not be converted to a ``++std::string++`` when calling the function ``++find++`` (this function is now a member function template). However, each time during the search when a comparison will be needed between ``++MyString++`` and ``++std::string++``, a conversion will now happen to convert ``++str++`` to always the same ``++string++``. A unique conversion is replaced by ``++O(ln N)++`` conversions.

This problem appears also for the case of an unordered container for which heterogeneous lookup is enabled (available since {cpp}20), for which equality is used to compare elements in the same bucket (having same or close hashes):
----
void g() {
  std::unordered_map<std::string, int, StringHash, std::equal_to<>> m;
  MyString str{"abc"};
  m.find(str);
}
----

> Again in this case ``++str++`` will not be converted to a ``++std::string++`` when calling the function ``++find++``, and for each element of the bucket that corresponds to the hash of ``++str++`` will be compared using homogeneous ``++operator==++``,

Reviewer: ```suggestion
In this case ``++str++`` will not be converted to a ``++std::string++`` when calling the function ``++find++``, and each element of the bucket that corresponds to the hash of ``++str++`` will be compared using homogeneous ``++operator==++``,
```

> and for each such comparison, a conversion will now happen. The number of elements varies depending on the hash distribution from ``++O(1)++` (in average case), to  ``++O(N)++`` in the worst case.

Reviewer: ```suggestion
and for each such comparison, a conversion will now happen. The number of compared elements varies depending on the hash distribution from ``++O(1)++` (on average) to ``++O(N)++`` (in the worst case).
```

> As consequence, the performance of slow runs (where multiple hash collisions happen due the input data), is made even worse.

Reviewer: ```suggestion
As consequence, the performance of slow runs (when multiple hash collisions happen due to the data distribution) is made even worse.
```

  MyString str{"abc"}; // See previous definition of MyString
  m.find(str); // Compliant, one conversion at the start
}

void g() {
  std::unordered_map<std::string, int> m;
  MyString str{"abc"}; // See previous definition of MyString
  m.find(str); // Compliant, one conversion at the start
}

----
Option 2: Provide heterogeneous comparisons

----

> bool operator==(MyString const &s1, std::string const &s2) {/*...*/} // automatically reversed from C++20

Reviewer: What do you mean by "automatically reversed from C++20"?

NB: east-const is used on these lines. Maybe this should be updated too?

Author: That you do need the `operator==(std::string const&, MyString const&)`, as the language will reverse order of arguments automatically
